### ISSUE TITLE
Feature request: Standardize `CLASS` definition

### ISSUE THREAD

**Issue by ConjuringCoffee** [SAP/abap-cleaner#157]
Hi Jörg-Michael, I'd like to request a new rule for standardizing the definition of a class (everything between `CLASS` and the period) concerning line-breaks, indentation and the order of the key words.

Personally, I don't think I have strong feelings on how it should like exactly. I'd just like it to be consistent. 😊 Thank you!

Before:
```abap
CLASS ycl_test DEFINITION
  FINAL
  PUBLIC FOR TESTING
  CREATE PUBLIC INHERITING FROM ycl_test_2.
```

Example for after:
```abap
CLASS ycl_test DEFINITION
  PUBLIC
  CREATE PUBLIC
  FINAL
  FOR TESTING
  INHERITING FROM ycl_bgi_function_toggles.
```

**Comment by jmgrassau** [comment]
Hi ConjuringCoffee, 

couldn't agree more! I think the order of the additions should be as shown in the ABAP documentation (nobody can remember this order, so perfect case for automated standardization :-) …

```
... [PUBLIC] 
    [INHERITING FROM superclass] 
    [ABSTRACT] 
    [FINAL] 
    [CREATE {PUBLIC|PROTECTED|PRIVATE}] 
    [SHARED MEMORY ENABLED] 
    [FOR TESTING] 
    [FOR BEHAVIOR OF] 
    [GLOBAL] FRIENDS class1 class2 ... 
                      intf1  intf2  ...].
```

… but you'd maybe want a few options (happy to get suggestions) on 
* whether to continue after "CLASS … DEFINITION" (at least if everything fits into one line?)
* whether to group some of those additions on the same line – e.g. without inheritance, you could maybe have those very common additions "PUBLIC FINAL CREATE PRIVATE" in one line(?), or the whole "FOR TESTING RISK LEVEL … DURATION …"  addition,
* how to indent them – I personally like indenting them below the class name, because for me, indent+2 indicates a new statement in an inner block, while indent+(4 or more) indicates that the same statement continues.

Kind regards,
Jörg-Michael

**Comment by ConjuringCoffee** [comment]
It's great to see that my suggestion was so well received!

> whether to continue after "CLASS … DEFINITION" (at least if everything fits into one line?)

Hmm, I sometimes do this, but I'm not following any strict rules so far. I think for local classes, I put everything into one line if they have no inheritance. 

```abap
CLASS lcl_test_data DEFINITION CREATE PRIVATE FOR TESTING.
```

When inheritice is used, I like to put that on a separate line:

```abap
CLASS lcx_no_data_found DEFINITION FINAL FOR TESTING
  INHERITING FROM zcx_static_check.
ENDCLASS.
```

For global classes, I generally put each key word on a separate line (probably because that's how it is set up when the class is initially generated).

> whether to group some of those additions on the same line – e.g. without inheritance, you could maybe have those very common additions "PUBLIC FINAL CREATE PRIVATE" in one line(?), or the whole "FOR TESTING RISK LEVEL … DURATION …" addition,

I'm not sure. I know that I'd want "FOR TESTING RISK LEVEL … DURATION …" on separate lines though. 😊

> how to indent them – I personally like indenting them below the class name, because for me, indent+2 indicates a new statement in an inner block, while indent+(4 or more) indicates that the same statement continues.

I personally like to indent them with only two spaces. Your argument is sound, but it would look very different to the rest of the system. Indentation with two spaces for the class definition "is just what classes look like"... (Oh no, am I getting old? 😥)

**Comment by jmgrassau** [comment]
Hi ConjuringCoffee, 

great, thanks for the feedback! 

> two spaces for the class definition "is just what classes look like"... (Oh no, am I getting old? 😥)

😂 I promise to make two spaces at least part of what _can_ be configured ;-)

Kind regards,
Jörg-Michael

**Comment by jmgrassau** [comment]
Hi ConjuringCoffee, 

thanks again for this idea! This new rule "**Standardize CLASS ... DEFINITION**" is now available with version 1.9.0, which was just released. The rule puts everything to the correct order, as specified by the ABAP Keyword Documentation. The big question is where and when to put line breaks. For this, the rule does come with quite a number of options, but I think they are rather straightforward:

![image](https://github.com/SAP/abap-cleaner/assets/79639298/d9806918-3ddc-4812-bbf1-cee53387e0a2)

Indentation allows for 4 options:

![image](https://github.com/SAP/abap-cleaner/assets/79639298/e173b965-75dd-4e60-9518-44340e6f6454)

And one-liners can be created / kept / split as usual, however you can also opt for "two-liners" (i.e. CLASS ... DEFINITION in the first line, all the options in the next line):

![image](https://github.com/SAP/abap-cleaner/assets/79639298/8793f038-c0bf-4e88-ae90-b9174236712d)

Kind regards,
Jörg-Michael

**Comment by ConjuringCoffee** [comment]
Awesome, I'll look forward to using this 😄
